// express-orm/src/index.js
const mysql = require('mysql')
require('dotenv').config()

//import modules
const tableBuilder = require('./tableBuilder/index')
const insertTable = require('./insertTable/index')

class Connection {
    constructor(object){
        this.connection = []
        this.tables = []
        try{
            this.ENVIRONMENT = process.env.ENVIRONMENT
        }
        catch(err){
            this.ENVIRONMENT = null
        }
    }

    createConnection(name, object){
        if(!name) throw "no name defined"
        if(this.connection[name]) throw "connection already exists"
        if(this.environment == name){
            let connection = mysql.createPool(object)
            this.connection[name] = connection
        }
    }
    SQLquery(query, args = null){
        if(!args){
            args = null
        }
        if(!query) throw "no query defined"
        return new Promise(data => {
            this.connection[this.environment].query(query, args, (error, result, fields) => {
                if(error) throw error
                data(result)
            })
        })
    }
    createTable(name, object){
        let table = tableBuilder(name, object)
        this.tables[name] = object
        this.SQLquery(table)
        //.then(res => console.log(res))
        .catch(err => console.log(err))
        
    }
    setEnvironment(value){
        this.environment = value
    }
    getTableSchema(name){
        return this.tables[name]
    }
    Insert(table, values){
        try{
            return new Promise(data => {
                let result = insertTable(table, this.getTableSchema(table), values)
                if(result.status){
                    this.SQLquery(result.data.sql, result.data.valuelist)
                    .then(res => {
                        data(res)
                    })
                    .catch(err => {
                        throw err
                    })
                }
                else{
                    throw result.data
                }
            })
        }
        catch(err){
            console.log(err)
        }
    }
}

module.exports = Connection
/*

const conn = new Connection()

conn.setEnvironment("dev")

conn.createConnection('dev', {
    host: "localhost",
    user: "root",
    password: 'root',
    port: 3306,
    database: "testtable"
})

conn.createConnection('production', {
    host: "localhost",
    user: "root",
    password: 'root',
    port: 3306,
    database: "testtable"
})

conn.createTable('usertable', {
    name: { type: String, notNull: true },
    username: { type: String, length: 64, notNull: true, unique: true },
    email: { type: String, length: 255, notNull: true, unique: true },
    password: { type: String, length: 128, notNull: true, cypher: "sha256" },
    profile_picture: { type: String, notNull: true, default: "test" }
})


conn.Insert('usertable', {
    name: "Alexander Gauss",
    username: "alexandergauss1",
    email: "alexander.gauss.vienna@gmail.com",
    password: "alexander2003",
})
.then(res => {
    console.log(res)
})
.catch(err => {
    console.log(err)
})

*/
